src/merkle/empty_nodes.rs

```diff
@@ -0,0 +1,467 @@
+use super::{Felt, Rpo256, RpoDigest, Vec, Word, WORD_SIZE, ZERO};
+
+// EMPTY NODES SUBTREES
+// ================================================================================================
+
+/// An empty sub-trees set.
+///
+/// Will provide precomputed node values for path computation for pre-defined depths.
+pub struct EmptyNodesSubtrees;
+
+impl EmptyNodesSubtrees {
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an empty subtrees set for depth `64`.
+    pub const fn empty_subtrees_64() -> [RpoDigest; 65] {
+        EMPTY_SUBTREES_64
+    }
+
+    /// Fetches a precomputed node for an empty sub-tree with maximum depth `64`.
+    pub const fn get_node_64(depth: u8) -> RpoDigest {
+        debug_assert!(depth <= 64);
+        EMPTY_SUBTREES_64[depth as usize]
+    }
+
+    /// Computes a sequence of empty hashes for an arbitrary depth
+    pub fn compute_empty_hashes(depth: u8) -> Vec<RpoDigest> {
+        (0..depth as u32 + 1)
+            .scan(Word::default().into(), |state, _| {
+                let value = *state;
+                *state = Rpo256::merge(&[value, value]);
+                Some(value)
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
+            .collect()
+    }
+}
```

Comment:
That wouldn't work if we have a depth smaller than `64` as the returned sub-tree wouldn't open to zero.

Every generated sub-tree is a set of its own. We can, alternatively, generate constants for all the depths, and pick the appropriate one for the passed argument.

Right now, we have only `64`, that is the case for the tiered sparse merkle tree.

Reply:
The principle behind the idea is very good. We need to make a couple of small modifications to that:

1) The function should be `const`
1) The returned vector must always open to zero

To satisfy `1`, we just recreate a slice from the static space that is a subset of the constant. I just wonder if we should `assert` instead of `debug_assert` as this would be a production requirement to not return invalid slices. To prevent that, I added zeroes to the remainder of the constant until the maximum possible `u8` (that is 255), so if the user prompts for a depth larger than 64, he will get an invalid subtree, but not an invalid memory reference.

And for `2`, the slice must have its last element on `64`, so we just use a range `64 - depth..65`

Reply:
I'd put either `assert` or `debug_assert` to check if the depth is > 64. But if this breaks the ability to do `const fn` - I'd keep the current implementation.

Reply:
Added an `assert` and truncated the constant.

I added a `Panic` documentation for the function, but we should probably have a better way to document panic-on-input functions so we can easily safeguard our stack before we move these to production